### PR TITLE
Feature/add s3 write

### PR DESCRIPTION
**Pull request by BrianKolowitz**
Hi, we wanted to do something like ./run_synthea -p 100000000 and skip local file storage so we added an s3 writer. the code is now extensible and can add other writers such as an HTTPWriter which could post direct to a FHIR server. The PR was getting pretty big  so we stopped just short of doing that.

### CODE REVIEW

File: src/main/java/org/mitre/synthea/export/FhirPractitionerExporterDstu2.java [hunk 1]
```diff
@@ -65,19 +64,21 @@ public static void export(long stop) {
...
-        Files.write(outFilePath, Collections.singleton(bundleJson), StandardOpenOption.CREATE_NEW);
+        // get output folder
+        List<String> folders = new ArrayList<>();
+        folders.add("fhir_dstu2");
+        String baseDirectory = Config.get("exporter.baseDirectory");
+        File f = Paths.get(baseDirectory, folders.toArray(new String[0])).toFile();
+        f.mkdirs();
+        Path outFilePath = f.toPath().resolve("practitionerInformation" + stop + ".json");
+        if (Boolean.parseBoolean(Config.get("exporter.upload_directly_to_aws_s3"))) {
+          AWSS3Writer.appendToFile("fhir_stu3", "practitionerInformation" + stop + ".json", Collections.singleton(bundleJson).toString());
```
Comment (jawalonoski) [inline src/main/java/org/mitre/synthea/export/FhirPractitionerExporterDstu2.java]: `fhir_stu3` should be `fhir_dstu2`

File: src/main/java/org/mitre/synthea/export/FhirPractitionerExporterR4.java [hunk 2]
```diff
@@ -54,28 +53,28 @@ public static void export(long stop) {
...
-        Files.write(outFilePath, Collections.singleton(bundleJson), StandardOpenOption.CREATE_NEW);
+        // get output folder
+        List<String> folders = new ArrayList<>();
+        folders.add("fhir");
+        String baseDirectory = Config.get("exporter.baseDirectory");
+        File f = Paths.get(baseDirectory, folders.toArray(new String[0])).toFile();
+        f.mkdirs();
+        Path outFilePath = f.toPath().resolve("practitionerInformation" + stop + ".json");
+        if (Boolean.parseBoolean(Config.get("exporter.upload_directly_to_aws_s3"))) {
+          AWSS3Writer.appendToFile("fhir_stu3", "practitionerInformation" + stop + ".json", Collections.singleton(bundleJson).toString());
```
Comment (jawalonoski) [inline src/main/java/org/mitre/synthea/export/FhirPractitionerExporterR4.java]: `fhir_stu3` should be `fhir_r4`

File: src/main/java/org/mitre/synthea/export/HospitalExporterR4.java [hunk 3]
```diff
@@ -42,27 +43,28 @@ public static void export(long stop) {
...
-        Files.write(outFilePath, Collections.singleton(bundleJson), StandardOpenOption.CREATE_NEW);
+        // get output folder
+        List<String> folders = new ArrayList<>();
+        folders.add("fhir");
+        String baseDirectory = Config.get("exporter.baseDirectory");
+        File f = Paths.get(baseDirectory, folders.toArray(new String[0])).toFile();
+        f.mkdirs();
+        Path outFilePath = f.toPath().resolve("hospitalInformation" + stop + ".json");
+        if (Boolean.parseBoolean(Config.get("exporter.upload_directly_to_aws_s3"))) {
+          AWSS3Writer.appendToFile("fhir_stu3", "practitionerInformation" + stop + ".json", Collections.singleton(bundleJson).toString());
```
Comment (jawalonoski) [inline src/main/java/org/mitre/synthea/export/HospitalExporterR4.java]: `fhir_stu3` should be `fhir_r4`

File: src/main/java/org/mitre/synthea/export/HospitalExporterDstu2.java [hunk 4]
```diff
@@ -42,27 +43,28 @@ public static void export(long stop) {
...
-        Files.write(outFilePath, Collections.singleton(bundleJson), StandardOpenOption.CREATE_NEW);
+        // get output folder
+        List<String> folders = new ArrayList<>();
+        folders.add("fhir_dstu2");
+        String baseDirectory = Config.get("exporter.baseDirectory");
+        File f = Paths.get(baseDirectory, folders.toArray(new String[0])).toFile();
+        f.mkdirs();
+        Path outFilePath = f.toPath().resolve("hospitalInformation" + stop + ".json");
+        if (Boolean.parseBoolean(Config.get("exporter.upload_directly_to_aws_s3"))) {
+          AWSS3Writer.appendToFile("fhir_stu3", "practitionerInformation" + stop + ".json", Collections.singleton(bundleJson).toString());
```
Comment (jawalonoski) [inline src/main/java/org/mitre/synthea/export/HospitalExporterDstu2.java]: `fhir_stu3` should be `fhir_dstu2`

File: .vscode/settings.json [hunk 5]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (jawalonoski) [inline .vscode/settings.json]: You should not include any editor settings in the repo. Please remove `.vscode/settings.json`

File: src/main/java/org/mitre/synthea/export/Exporter.java [hunk 6]
```diff
@@ -8,17 +8,15 @@
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
-import java.nio.file.StandardOpenOption;
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.Iterator;
-import java.util.List;
+import java.util.*;
```
Comment (jawalonoski) [inline src/main/java/org/mitre/synthea/export/Exporter.java]: No star imports, please.

File: src/main/java/org/mitre/synthea/export/Exporter.java [hunk 7]
```diff
@@ -418,52 +432,21 @@ private static boolean entryWithinTimeRange(
...
    * @param person    The person being exported.
-   * @param tag       A tag to add to the filename before the extension.
+   * @param tag       A tag to add to the fileName before the extension.
    * @param extension The file extension to use.
-   * @return The filename only (not a path).
+   * @return The fileName only (not a path).
    */
   public static String filename(Person person, String tag, String extension) {
-    if (Boolean.parseBoolean(Config.get("exporter.use_uuid_filenames"))) {
+    if (Boolean.parseBoolean(Config.get("exporter.use_uuid_fileNames"))) {
```
Comment (jawalonoski) [inline src/main/java/org/mitre/synthea/export/Exporter.java]: You changed the configuration setting name without changing the configuration file.

The configuration settings do not use camel-case, please revert to `exporter.use_uuid_filenames`

Comment (BrianKolowitz) [reply src/main/java/org/mitre/synthea/export/Exporter.java]: fixed

File: src/main/java/org/mitre/synthea/export/Exporter.java (hunk 6 above)
Comment (BrianKolowitz) [reply src/main/java/org/mitre/synthea/export/Exporter.java]: fixed

**Comment by BrianKolowitz** [comment]
I did not run ```./gradlew check``` I'll make sure to do that going forward

File: src/main/java/org/mitre/synthea/export/FhirPractitionerExporterDstu2.java (hunk 1 above)
Comment (BrianKolowitz) [reply src/main/java/org/mitre/synthea/export/FhirPractitionerExporterDstu2.java]: fixed

File: src/main/java/org/mitre/synthea/export/FhirPractitionerExporterR4.java (hunk 2 above)
Comment (BrianKolowitz) [reply src/main/java/org/mitre/synthea/export/FhirPractitionerExporterR4.java]: fixed

File: src/main/java/org/mitre/synthea/export/HospitalExporterR4.java (hunk 3 above)
Comment (BrianKolowitz) [reply src/main/java/org/mitre/synthea/export/HospitalExporterR4.java]: fixed

File: src/main/java/org/mitre/synthea/export/HospitalExporterDstu2.java (hunk 4 above)
Comment (BrianKolowitz) [reply src/main/java/org/mitre/synthea/export/HospitalExporterDstu2.java]: fixed

File: .vscode/settings.json (hunk 5 above)
Comment (BrianKolowitz) [reply .vscode/settings.json]: fixed

**Comment by dehall** [comment]
Recently a PR was rejected as an "external systems integration" and this note was added to the wiki:  https://github.com/synthetichealth/synthea/wiki/Contributing#contributions-to-avoid 

Would it be possible to make this PR a little more generic so that it's easier to plug in different cloud providers, or a FHIR server as mentioned? (note that SyntheticMass is now hosted on Google Cloud)

In terms of the specific S3 integration, I have no strong feelings one way or the other but I think the philosophy of avoiding specific external integrations makes sense. If we merge one then we are on the hook for maintaining it. But if a generic API is in place then people can easily add their own.

**Comment by jawalonoski** [comment]
@BrianKolowitz or @JeffreyTaylor -- any thoughts on @dehall questions?

**Comment by irjudson** [comment]
@jawalonoski I've created a branch to add jclouds storage as an exporter. Is that what you were thinking? If so, I'm curious your thoughts on baseDirectory config -- it's applicable in any location, but the implicit assumption that there's a (local) filesystem in the export hierarchy feels like something will bite me as I work through this.

I'll make PR's for code review in a separate branch if that's ok?

**Comment by shabiel** [comment]
https://github.com/synthetichealth/synthea/wiki/Contributing

Esp:

External Systems Integrations

We don't typically accept pull requests of this type -- we consider
the post-processing, extract, transformation, and loading (ETL) of
Synthea data into external systems to be outside the scope of the
project. While loading data into your system of choice important to
you, it may not be to others, who might prefer OMOP, VistA, Document
DBs (e.g. Mongo DB), Elastic Search, or something else entirely.

While we aim to support multiple formats (e.g. FHIR, C-CDA, CSV), we
do not intend to directly support external systems integrations.
Please do not issue a pull request for external systems integrations.

**Comment by irjudson** [comment]
Thank you @shabiel -- this must have been added to the wiki after the initial conversation here. It would be super helpful to keep the reference from here to there (so others don't assume the suggestion to make it generic is more important than the exclusivity of external systems).

Also, maybe just close this PR so it's not fishing for support?

I will support the external system in a fork I maintain because generating huge quantities of data for local use has limited value and ETL'ing is kind of monotonous when it can be done automatically.

Again, thanks for the response.

**Comment by jawalonoski** [comment]
Thank you for the pull request, but it is being closed due to conflicts and inactivity.

**Pull request closed without merging**